### PR TITLE
[Performance] Speed up Azure A10 instance creation

### PR DESCRIPTION
**Pull request by yika-luo**
Use the new custom image for Azure's A10 VM creation. 

Example performance change (with fixed region):
|  VM Type 💻   | Old Provision 🕐                 | New Provision 🕐     | % Speedup ✅ |
| --------------- | ---------------------------- | ----------- | ----------- |
| Azure GPU      |   21min    | 2min 50s   | 85% (7x) |

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual test: `time sky launch --cloud=azure -y --gpus=A10:1` and ssh to machine to confirm NVIDIA and nvidia-docker runtime installation

### CODE REVIEW

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+sudo apt update
+sudo apt install -y build-essential
+
+echo "Installing GRID driver..."
+GRID_DRIVER_URL="https://download.microsoft.com/download/8/d/a/8da4fb8e-3a9b-4e6a-bc9a-72ff64d7a13c/NVIDIA-Linux-x86_64-535.161.08-grid-azure.run"
+GRID_DRIVER_FILE="NVIDIA-Linux-x86_64-535.161.08-grid-azure.run"
+
+wget -nv $GRID_DRIVER_URL -O $GRID_DRIVER_FILE
+sudo chmod +x $GRID_DRIVER_FILE
+sudo sh $GRID_DRIVER_FILE --silent
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:12]:
```suggestion
sudo sh $GRID_DRIVER_FILE --silent --disable-nouveau
```
Should we disable nouveau driver? It seems there will be a significant performance difference for oss vs enterprise driver.

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:13]: There are a few additional steps to follow in the instruction here: https://learn.microsoft.com/en-us/azure/virtual-machines/linux/n-series-driver-setup#ubuntu-1

Should we add those as well, including setting up several files? That might relates why we are having bad performance when the extension is specified.

File: sky/provision/azure/instance.py [hunk 3]
```diff
@@ -313,27 +313,6 @@ def _create_vm(
     )
     # poller.result() will block on async operation until it's done.
     logger.info(f'Created VM {vm_poller.result().name}.')
```
Comment (Michaelvll) [inline sky/provision/azure/instance.py:315]: Can we have an additional line for waiting for the poller? It is mainly for readability, logging is mostly for non-stateful operations.

**Review by Michaelvll — COMMENTED** [review]
Nice! Thanks @yika-luo! This is a significant speed up! Left a few comments

File: sky/clouds/azure.py [hunk 4]
```diff
@@ -350,10 +354,6 @@ def make_deploy_resources_variables(
                 'image_version': version,
             }
 
-        # Setup the A10 nvidia driver.
-        need_nvidia_driver_extension = (acc_dict is not None and
-                                        'A10' in acc_dict)
-
```
[the remark this answers is not in the document]
Comment (yika-luo) [reply sky/clouds/azure.py:356]: Discussed offline

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh (hunk 2 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:13]: Should we run step 7-9 in the reference above?

File: sky/provision/azure/instance.py [hunk 5]
```diff
@@ -311,29 +311,8 @@ def _create_vm(
         vm_name=vm_name,
         parameters=vm_instance,
     )
-    # poller.result() will block on async operation until it's done.
+    vm_poller.wait()
     logger.info(f'Created VM {vm_poller.result().name}.')
```
Comment (Michaelvll) [inline sky/provision/azure/instance.py:315]: Just curious, why don't set the result to a variable and reuse them later?
```suggestion
    # Wait for the creation of VM and get the result.
    vm_poller_result = vm_poller.result()
    logger.info(f'Created VM {vm_poller_result.name}.')
```

**Review by Michaelvll — COMMENTED** [review]
Thanks @yika-luo! Left a comment below, why don't we run step 7-9 as suggested in the Azure grid driver installer doc?

File: sky/provision/azure/instance.py (hunk 5 above)
Comment (yika-luo) [reply sky/provision/azure/instance.py]: No strong reasons, just having less code

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh [hunk 6]
```diff
@@ -11,6 +11,11 @@ wget -nv $GRID_DRIVER_URL -O $GRID_DRIVER_FILE
 sudo chmod +x $GRID_DRIVER_FILE
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:11]: nit: it seems we have two installation script in the home dir on the VM created with this image. let's remove them:
```bash
NVIDIA-Linux-x86_64-535.161.08-grid-azure.run  cuda_12.2.0_535.54.03_linux.run
```

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh [hunk 7]
```diff
@@ -11,6 +11,11 @@ wget -nv $GRID_DRIVER_URL -O $GRID_DRIVER_FILE
...
 sudo sh $GRID_DRIVER_FILE --silent --disable-nouveau
 
+echo "Set vGPU Licensing Daemon config..."
+sudo cp /etc/nvidia/gridd.conf.template /etc/nvidia/gridd.conf
+sudo sed -i '/^FeatureType=0/s/^/# /' /etc/nvidia/gridd.conf
+echo "IgnoreSP=FALSE" | sudo tee -a /etc/nvidia/gridd.conf
+echo "EnableUI=FALSE" | sudo tee -a /etc/nvidia/gridd.conf
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:18]: I am not sure why, but when I create an instance with the `sky launch -c test-azure-a10 --cloud azure --gpus A10 --image-id /CommunityGalleries/skypilot-5c783dcb-c1d8-48ca-a5eb-0cee59e165fd/Images/skypilot-gpu-gen2-grid`. The `/etc/nvidia/gridd.conf` file only has the following. Is that expected?
```
EnableUI=FALSE
```

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+echo "Installing CUDA toolkit..."
+CUDA_TOOLKIT_URL="https://developer.download.nvidia.com/compute/cuda/12.2.0/local_installers/cuda_12.2.0_535.54.03_linux.run"
+CUDA_TOOLKIT_FILE="cuda_12.2.0_535.54.03_linux.run"
+wget -nv $CUDA_TOOLKIT_URL -O $CUDA_TOOLKIT_FILE
+sudo sh $CUDA_TOOLKIT_FILE --silent --toolkit --override
+
+# Set environment variables
+echo 'export PATH=$PATH:/usr/local/cuda-12.2/bin' >> ~/.bashrc
+echo 'export LD_LIBRARY_PATH=/usr/local/cuda-12.2/lib64' >> ~/.bashrc
```
Comment (Michaelvll) [inline sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:28]: Are we discarding all other LD_LIBRARY_PATH? That does not feel correct to me.
```suggestion
echo 'export LD_LIBRARY_PATH=/usr/local/cuda-12.2/lib64' >> ~/.bashrc
```

**Comment by Michaelvll** [comment]
A future TODO: we need to add a smoke test for A10 on Azure

File: sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh (hunk 7 above)
Comment (yika-luo) [reply sky/clouds/service_catalog/images/provisioners/cuda-azure-grid.sh:18]: I regenerated the image so things should look right now

**Review by Michaelvll — APPROVED** [review]
LGTM. Thanks @yika-luo!

**Pull request merged**